packages/context/src/binding.ts

```diff
@@ -162,9 +166,9 @@ export class Binding<T = BoundValue> {
 
   private _cache: WeakMap<Context, T>;
   private _getValue: (
-    ctx?: Context,
-    session?: ResolutionSession,
-  ) => ValueOrPromise<T>;
+    ctx: Context,
+    optionsOrSession?: ResolutionOptionsOrSession,
+  ) => ValueOrPromise<T | undefined>;
```

Comment:
I am concerned about this change, we don't want to start leaking `undefined` to code that was expecting a `T` value instead. I'll take a closer look tomorrow.

Reply:
Please note `_getValue` is an internal method.

Reply:
This change is needed to allow an alias binding to resolve to the `undefined` with `optional: true` even when the target binding does not exist.

Reply:
> This change is needed to allow an alias binding to resolve to the `undefined` with `optional: true` even when the target binding does not exist.

Thanks for explaining this scenario, I agree it's something we need to take care of.

So far, our type definitions for Context and Binding methods went a long way to allow the compiler to catch programming errors related to `undefined` values.

Cross-posting https://github.com/strongloop/loopback-next/pull/1169#issuecomment-375362267

> Consider the following REST binding:
> 
> ```ts
> export const HOST = new BindingKey<string | undefined>('rest.host');
> ```
>
> Before my change, there was no indication that undefined is a valid value for the HOST binding. Without that knowledge, one could write following code:
>
> ```ts
> const resolve = promisify(dns.resolve);
>
> const host = await ctx.get<string>(RestBindings.HOST);
> const records = await resolve(host);
> // etc.
> ```
> 
> The compiler will happily accept such code and only at runtime we will learn that HOST may be also undefined in which case our code needs to find the server's host name using a different way.

Now what I would like to avoid is the situation when an alias binding is defined with value of type `T`, but the actual runtime type is `T | undefined`.

Let me review the actual code changes to see if this concern is relevant or not.

Reply:
Hey, binding aliases are not part of this pull request! Let's revert this change please and discuss it in #2635 as part of `toAlias` implementation.

Reply:
I refactored it into a separate commit. Please also note that `Binding.getValue()` still returns `ValueOrPromise<T>`. Only the `_getValue` is changed.